ecommerce/backend/API/views/search.py

```diff
@@ -5,22 +5,76 @@
 from rest_framework.renderers import JSONRenderer
 
 from API.utils import permissions, Role
-from API.models import Product, Subcategory, Vendor, Brand
+from API.models import Product, ProductIndex, Subcategory, Vendor, Brand
 from django.db.models import Q
 #from API.serializers import search_serializer
 from API.serializers.product_serializer import ProductResponseSerializer
 from API.serializers.vendor_serializer import VendorResponseSerializer
 
+import string
+import urllib.request
+import ssl
+import json
+import requests
+
+API_URL = "https://api.datamuse.com/words?ml="
+
+# these are the stopwords of the NLTK library, taken from https://gist.github.com/sebleier/554280
+with open('stopwords.txt', 'r', encoding='latin-1') as f:   
+    stopwords = [line.strip() for line in f.readlines()]
+
+# preprocess the sentence by applyin case-folding, punctutation removal and tokenization
+def preprocess(sentence):
+    global stopwords
+    sentence = sentence.casefold() # case-folding
+    sentence = sentence.translate(str.maketrans(string.punctuation, ' '*len(string.punctuation))) #punctuation removal
+    sentence = sentence.split() #tokenization
+    sentence = [word for word in sentence if not word in stopwords] # stopword removal
+    return sentence
+
+# converts the search-related fields into one preprocessed text
+def get_indexed_text_of_product(product):
+    data = ProductResponseSerializer(product).data
+    text = " ".join(preprocess(" ".join([data['name'], data['long_description'],
+        data['short_description'], data['category']['name'], data['subcategory']['name'],
+            data['vendor']['name']])))
+    return text
+
+# gets similar queries from the Datamuse API
+def get_similar_queries(query):
+    processed_query = '+'.join(preprocess(query))
+    # get the content of the API
+    r = requests.get(API_URL+processed_query)
+    results = r.content
+    results = json.loads(results)
+    # sort wrt to the highest score which means semantically more similar
+    results = sorted(results, key= lambda x:x['score'], reverse=True)
+    # return words of the top 10 results
+    similar_queries = [d['word'] for d in results[:10]]
+    return similar_queries
```

Comment:
Excellent work but this function should not represented in search.py

Reply:
Aggred. It is better to place it in utils

Reply:
Functions are moved into utils/semantic_search.py